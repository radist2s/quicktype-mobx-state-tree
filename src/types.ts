import {types as mstTypes} from 'mobx-state-tree'

function isNumericString(str?: string | null) {
    return typeof str === 'string' && /^[\d\.]+$/.test(str)
}

/**
 * Returns timestamp in milliseconds from parsed input literal or numeric seconds timestamp
 * @param timestamp
 */
function normalizeTimestampMs(timestamp: string | number | null): number | null {
    if (typeof timestamp === 'string') {
        if (isNumericString(timestamp)) {
            return parseFloat(timestamp) * 1000
        }

        return Date.parse(timestamp)
    } else if (typeof timestamp === 'number') {
        return timestamp * 1000
    }

    return null
}

type DateSnapshotType = string | number | null
type DateType = Date | null

export function typeDate() {
    return mstTypes.custom<DateSnapshotType, DateType>({
        name: 'Date',
        fromSnapshot(snapshot: DateSnapshotType): DateType {
            const timestampMs = normalizeTimestampMs(snapshot)

            return timestampMs === null ? null : new Date(timestampMs)
        },
        getValidationMessage(snapshot: string | null): string {
            const timestampMs = normalizeTimestampMs(snapshot)

            const isValidTimestamp = timestampMs === null || isFinite(timestampMs)

            return isValidTimestamp ? '' : `'${snapshot}' doesn't look like a valid literal timestamp or unix timestamp`
        },
        isTargetType(value: DateSnapshotType | DateType): boolean {
            return value instanceof Date || value === null
        },
        toSnapshot(value: DateType): Exclude<DateSnapshotType, number> {
            return value ? value.toISOString() : null
        }
    })
}

const {Date: MstDate, ...mstPartialProps} = mstTypes

export const types = {
    ...mstPartialProps,
    get Date() {
        return typeDate()
    }
}
