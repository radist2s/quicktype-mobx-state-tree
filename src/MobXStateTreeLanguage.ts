import {
    TypeScriptRenderer,
    TypeScriptTargetLanguage,
    RenderContext,
    getOptionValues,
    tsFlowOptions,
    ClassType,
    Name,
    Type,
    modifySource,
    ClassProperty,
    Sourcelike,
    matchType,
    TargetLanguage,
    panic,
    nullableFromUnion,
    defined,
    Option
} from 'quicktype/dist/quicktype-core'
import {utf16StringEscape} from 'quicktype/dist/quicktype-core/support/Strings'
import {isES3IdentifierStart} from 'quicktype/dist/quicktype-core/language/JavaScriptUnicodeMaps'
import {legalizeName} from 'quicktype/dist/quicktype-core/language/JavaScript'
import {MultiWord, singleWord, parenIfNeeded} from 'quicktype/dist/quicktype-core/Source'
import {OptionValues, BooleanOption} from 'quicktype/dist/quicktype-core/RendererOptions'
import {TypeKind} from 'quicktype/dist/quicktype-core/Type'
import {ConvertersOptions} from 'quicktype/dist/quicktype-core/support/Converters'

export const mstOptions =  {
    runtimeConverterTypes: new BooleanOption("runtime-type-converters", "Normalize string data such 'Date' to valid MST snapshot format at runtime", true),
    nicePropertyNames: tsFlowOptions.nicePropertyNames,
    acronymStyle: tsFlowOptions.acronymStyle,
    converters: tsFlowOptions.converters
} as const

export class MobXStateTreeTargetLanguage extends TypeScriptTargetLanguage {
    constructor(
        readonly displayName: string = 'MobX-State-Tree',
        readonly names: string[] = ['mobx', 'mst', 'ts', 'tsx', 'js', 'jsx'],
        readonly extension: string = 'ts'
    ) {
        super()
    }

    protected makeRenderer(
        renderContext: RenderContext,
        untypedOptionValues: {[name: string]: any}
    ): MobXStateTreeRenderer {
        untypedOptionValues = Object.assign({}, untypedOptionValues, {
            // [tsFlowOptions.justTypes.definition.name]: String(true),
            [tsFlowOptions.declareUnions.definition.name]: String(false), // set to false, MST has builtin support fot union types
            [tsFlowOptions.runtimeTypecheck.definition.name]: String(false) // set to false, MST has builtin support for type checking
        })

        const typedOptions = {
            ...getOptionValues(mstOptions, untypedOptionValues)
        }

        return new MobXStateTreeRenderer(this, renderContext, typedOptions)
    }

    protected getOptions(): Option<any>[] {
        return Object.values(mstOptions)
    }
}

function quotePropertyName(original: string): string {
    const escaped = utf16StringEscape(original)
    const quoted = `"${escaped}"`

    if (original.length === 0) {
        return quoted
    } else if (!isES3IdentifierStart(original.codePointAt(0) as number)) {
        return quoted
    } else if (escaped !== original) {
        return quoted
    } else if (legalizeName(original) !== original) {
        return quoted
    } else {
        return original
    }
}

export class MobXStateTreeRenderer extends TypeScriptRenderer {
    private readonly _mstOptions: OptionValues<typeof mstOptions>

    constructor(
        targetLanguage: TargetLanguage,
        renderContext: RenderContext,
        options: OptionValues<typeof mstOptions>
    ) {
        const {runtimeConverterTypes, ...optionsRest} = options as OptionValues<(typeof mstOptions & typeof tsFlowOptions)>

        super(targetLanguage, renderContext, optionsRest)

        this._mstOptions = options
    }

    protected emitClassBlock(c: ClassType, className: Name): void {
        this.emitBlock(['export const ', className, ` = types.model('`, className, `', `], ')', () => {
            this.emitClassBlockBody(c)
        })
    }

    protected emitClassBlockBody(c: ClassType): void {
        this.emitPropertyTable(c, (name, _jsonName, p) => {
            const t = p.type
            return [
                [modifySource(quotePropertyName, name), ': ', p.isOptional ? 'types.maybe(' : ''],
                [this.sourceFor(t).source, p.isOptional ? ')' : '', ',']
            ]
        })
    }

    protected sourceFor(t: Type): MultiWord {
        if (['class', 'object'].indexOf(t.kind) >= 0) {
            return singleWord(this.nameForNamedType(t))
        }

        return matchType<MultiWord>(
            t,
            _anyType => singleWord('types.frozen()'),
            _nullType => singleWord('types.null'),
            _boolType => singleWord('types.boolean'),
            _integerType => singleWord('types.number'),
            _doubleType => singleWord('types.number'),
            _stringType => singleWord('types.string'),
            arrayType => {
                const itemType = this.sourceFor(arrayType.items)

                return singleWord(['types.array(', parenIfNeeded(itemType), ')'])
            },
            _classType => panic('We handled this above'),
            mapType => singleWord(['types.frozen()']),
            _enumType => {
                const itemTypeName = this.nameForNamedType(_enumType)

                return singleWord([
                    'types.enumeration<',
                    itemTypeName,
                    `>('`,
                    itemTypeName,
                    `', `,
                    `Object.values(`,
                    itemTypeName,
                    `))`
                ])
            },
            unionType => {
                if (true/*!this._mstOptions.declareUnions*/ || nullableFromUnion(unionType) !== null) {
                    const uniqueChildren = Array.from(unionType.getChildren()).filter(function (
                        filteringChild,
                        index,
                        children
                    ) {
                        if (!filteringChild.isPrimitive()) {
                            return true
                        }

                        return index == children.findIndex(findChildren => findChildren.kind === filteringChild.kind)
                    })

                    const childrenSeparated = uniqueChildren.reduce<Sourcelike[]>((carry, childType) => {
                        // Join union children with comma if not first children is preparing
                        if (carry.length) {
                            carry = carry.concat(', ')
                        }

                        // filter out same kind union child types
                        const childSource = this.sourceFor(childType)

                        if (carry.find((carryType: Sourcelike | Type) => carryType === childType)) {
                            return carry
                        }
                        return carry.concat(parenIfNeeded(childSource))
                    }, [])
                    return singleWord(['types.union(', ...childrenSeparated, ')'])
                } else {
                    return singleWord(this.nameForNamedType(unionType))
                }
            },
            transformedStringType => {
                if (transformedStringType.kind === 'date-time') {
                    return singleWord('types.Date')
                }

                return singleWord('string')
            }
        )
    }

    protected makeNameForTopLevel(t: Type, givenName: string, maybeNamedType: Type | undefined): Name {
        const givenNameOrder = 100
        const inferredOrder = 100
        return (this as any).makeNameForType(t, defined((this as any)._namedTypeNamer), givenNameOrder, inferredOrder)
    }

    protected makeNameForNamedType(t: Type): Name {
        const givenNameOrder = 10

        type OrderMap = Extract<TypeKind, 'object' | 'enum'>

        const orders: {[key in OrderMap]: number} = {
            object: 10,
            enum: 5
        }

        const inferredNameOrder = orders[t.kind as OrderMap] !== undefined ? orders[t.kind as OrderMap] : 30

        return (this as any).makeNameForType(
            t,
            defined((this as any)._namedTypeNamer),
            givenNameOrder,
            inferredNameOrder
        )
    }

    protected emitSourceStructure() {
        this.emitLine("import {types} from 'mobx-state-tree'")
        this.ensureBlankLine()
        super.emitSourceStructure()
    }

    protected emitTypes(): void {
        this.forEachNamedType(
            'leading-and-interposing',
            (c: ClassType, n: Name) => {
                this.emitDescription(this.descriptionForType(c))
                this.emitClassBlock(c, n)
            },
            (e, n) => this.emitEnum(e, n),
            (u, n) => this.emitUnion(u, n)
        )
    }

    protected get needsTypeDeclarationBeforeUse() {
        return true
    }

    protected canBeForwardDeclared(t: Type): boolean {
        return (['enum', 'object'] as TypeKind[]).indexOf(t.kind) !== -1
    }

    protected emitConvertModuleBody(): void {
        const converter = (t: Type, name: Name) => {
            // const typeMap = this.typeMapTypeFor(t)

            this.emitBlock([this.deserializerFunctionLine(t, name), ' '], '', () => {
                if (true/*!this._mstOptions.runtimeTypecheck*/) {
                    this.emitLine('return JSON.parse(json);')
                } else {
                    // this.emitLine('return cast(JSON.parse(json), ', typeMap, ');')
                }
            })

            this.ensureBlankLine()

            this.emitBlock([this.serializerFunctionLine(t, name), ' '], '', () => {
                if (true/*!this._mstOptions.runtimeTypecheck*/) {
                    this.emitLine('return JSON.stringify(value);')
                } else {
                    // this.emitLine('return JSON.stringify(uncast(value, ', typeMap, '), null, 2);')
                }
            })
        }

        switch (this._mstOptions.converters) {
            case ConvertersOptions.AllObjects:
                this.forEachObject('interposing', converter)
                break

            default:
                this.forEachTopLevel('interposing', converter)
                break
        }
    }
}
