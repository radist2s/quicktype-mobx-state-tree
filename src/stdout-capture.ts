let originalStdoutWrite: undefined | typeof process.stdout.write

export default async function captureStdOut<P>(callback: () => P | Promise<P>): Promise<[P, string] | []> {
    originalStdoutWrite = originalStdoutWrite || process.stdout.write.bind(process.stdout)

    let stdOutput = ''

    process.stdout.write = (chunk: Buffer | string): boolean => {
        if (typeof chunk === 'string') {
            stdOutput += chunk
        }

        return true
    }

    const stopCapture = function () {
        if (originalStdoutWrite) {
            process.stdout.write = originalStdoutWrite
        }

        return stdOutput
    }

    return [await callback(), stopCapture()]
}
