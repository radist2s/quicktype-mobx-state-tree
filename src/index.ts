#!/usr/bin/env node

import {relative as pathRelative, resolve as pathResolve, dirname as pathDirname, extname as pathExtname, join as pathJoin} from 'path'
import {statSync, lstatSync} from 'fs'
import {quicktype as quicktypeLib, quicktypeMultiFile} from 'quicktype/dist/quicktype-core'
import {CLIOptions, parseCLIOptions, makeQuicktypeOptions, writeOutput} from 'quicktype'
import {MobXStateTreeTargetLanguage, mstOptions} from './MobXStateTreeLanguage'
import {Options} from 'quicktype/dist/quicktype-core/Run'
import captureStdOut from './stdout-capture'

export async function quicktype(options: Partial<Options>) {
    const optionsMerged = Object.assign({}, options, {
        lang: new MobXStateTreeTargetLanguage()
    })

    const {lines: schema} = await quicktypeLib(optionsMerged)

    return schema.join('\n')
}

export function resolveTypesModuleOption(options: Partial<Options>, outPath?: string) {
    if (!options || !options.rendererOptions || !outPath || options.outputFilename === 'stdout') {
        return options
    }

    const {rendererOptions} = options

    const typesModule = mstOptions.typesModule.getValue(rendererOptions)

    if (!typesModule) {
        return options
    }

    const moduleAbsPath = pathResolve(typesModule)

    const moduleFileExtname = pathExtname(moduleAbsPath)

    const isFilePathModule = Array.prototype.some.call([moduleAbsPath], function (path) {
        try {
            const moduleLstat = statSync(path)

            return moduleLstat.isFile() || moduleLstat.isDirectory()
        } catch (e) {
            return false
        }
    })

    if (!isFilePathModule) {
        return options
    }

    const outDir: string = (() => {
        const outPathAbs = pathResolve(outPath)

        try {
            if (lstatSync(outPathAbs).isDirectory()) {
                return outPathAbs
            }
        } catch (e) {}

        return pathDirname(outPathAbs)
    })()

    const moduleImportPath: string = (() => {
        const moduleOutputPath = pathRelative(outDir, moduleAbsPath)

        return moduleFileExtname ? moduleOutputPath.slice(0, -moduleFileExtname.length) : moduleOutputPath
    })()

    return {
        ...options,
        rendererOptions: {
            ...rendererOptions,
            [mstOptions.typesModule.definition.name]:
                moduleImportPath[0] === '.' ? moduleImportPath : `./${moduleImportPath}`
        }
    }
}

export async function main(args: string[] | Partial<CLIOptions>) {
    const targetLanguage = new MobXStateTreeTargetLanguage()

    let cliOptions: CLIOptions
    if (Array.isArray(args)) {
        cliOptions = parseCLIOptions(args, targetLanguage)
    } else {
        throw new Error('Input arguments should be and array of cli options')
    }

    if (cliOptions.telemetry !== undefined) {
        switch (cliOptions.telemetry) {
            case 'enable':
                break
            case 'disable':
                break
            default:
                console.error("telemetry must be 'enable' or 'disable'")
                return
        }
        if (Array.isArray(args) && args.length === 2) {
            // This was merely a CLI run to set telemetry and we should not proceed
            return
        }
    }

    const [quicktypeOptions, stdOut] = await captureStdOut(async function () {
        const typedOptions = await makeQuicktypeOptions(cliOptions, [targetLanguage])
        if (!typedOptions) {
            return undefined
        }

        return resolveTypesModuleOption(typedOptions, cliOptions.out)
    })

    if (stdOut) {
        const packageJson = require(pathJoin(__dirname, '../package.json')) as {bin: object}
        const [[binName]] = Object.entries(packageJson.bin) // retrieves first binary property value from package.json file

        console.log(stdOut.replace(/\bquicktype\b/g, binName)) // replaces all all references to quicktype in help output with quicktypemst name
    }

    if (quicktypeOptions === undefined) {
        return
    }

    const resultsByFilename = await quicktypeMultiFile(quicktypeOptions)

    writeOutput(cliOptions, resultsByFilename)
}

if (require.main === module) {
    main(process.argv.slice(2)).catch(e => {
        if (e instanceof Error) {
            console.error(`Error: ${e.message}.`)
        } else {
            console.error(e)
        }
        process.exit(1)
    })
}
