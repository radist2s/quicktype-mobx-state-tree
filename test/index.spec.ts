import {InputData, JSONSchemaInput, JSONSchemaStore, JSONSchema} from 'quicktype/dist/quicktype-core'
import {join as pathJoin, sep} from 'path'
import {readFileSync, writeFileSync, unlinkSync} from 'fs'
import {quicktype} from '../src'
import {types, isUnionType, isType, isPrimitiveType} from 'mobx-state-tree'
import {
    IModelType,
    ModelPropertiesDeclarationToProperties,
    ModelPropertiesDeclaration
} from 'mobx-state-tree/dist/types/complex-types/model'

const schemaFilePath = pathJoin(__dirname, 'fixtures/product.schema.json')

class MyJSONSchemaStore extends JSONSchemaStore {
    async fetch(_address: string): Promise<JSONSchema | undefined> {
        return undefined
    }
}

async function getSchemaInout() {
    const schemaInput = new JSONSchemaInput(new MyJSONSchemaStore())

    const schemaName: string = 'Schema'

    // We could add multiple schemas for multiple types,
    // but here we're just making one type from JSON schema.
    await schemaInput.addSource({name: schemaName, schema: readFileSync(schemaFilePath).toString('utf-8')})

    const inputData = new InputData()

    inputData.addInput(schemaInput)

    return inputData
}

function normalizeDate(date: string | number): string | undefined {
    if (typeof date === 'string') {
        if ((/[^\d\s]+/).test(date)) {
            return `${Date.parse(date) / 1000}`
        }

        return date
    }
    else if (typeof date === 'number') {
        return String(date)
    }
}

function isDateType(type: IModelType<ModelPropertiesDeclarationToProperties<{}>, {}>) {
    return type && typeof type === 'object' && type.name === 'Date'
}

function snapshotDateNormalize<P extends ModelPropertiesDeclaration = {}>(snapshot: {[key in keyof P]: any}, model: IModelType<ModelPropertiesDeclarationToProperties<P>, {}>) {
    const snapshotPrepared = {...snapshot}

    for (const [property, value] of Object.entries(snapshot)) {
        const propertyType: any = model.properties[property] as any

        if (propertyType) {
            let isDatePropertyType = false

            if (isUnionType(propertyType)) {
                const subTypes = propertyType.getSubTypes() as any[]

                isDatePropertyType = subTypes && subTypes.find(isDateType)
            }
            else if (isDateType(propertyType)) {
                isDatePropertyType = true
            }

            if (isDatePropertyType) {
                snapshotPrepared[property as keyof P] = normalizeDate(value)
            }
        }
        else {
            delete snapshotPrepared[property]
        }
    }

    return snapshotPrepared
}

describe('transform', function () {
    it('should transform', async function () {
        expect(await quicktype({inputData: await getSchemaInout()})).toMatchSnapshot()
    })

    /*it('should load json snapshot', async function () {

        const modelSnapshot = require('./fixtures/schema-snapshot.json')

        const mstCode = await quicktype({inputData: await getSchemaInout()})

        const mstFilePath = `${__dirname}${sep}.${Math.floor(Math.random() * 100000)}.ts`

        const d = Product.properties

        let error: any

        try {
            writeFileSync(mstFilePath, mstCode)

            const mst = require(mstFilePath) as {Schema: IModelType<ModelPropertiesDeclarationToProperties<{}>, {}>}

            const Schema = types.snapshotProcessor(mst.Schema, {
                preProcessor(snapshot: any) {
                    const data = snapshotDateNormalize(snapshot, mst.Schema)

                    return data
                }
            })

            const schemaModel = Schema.create(modelSnapshot)

            console.log(schemaModel)
        }
        catch (e) {
            error = e
        }
        finally {
            unlinkSync(mstFilePath)

            if (error) {
                throw error
            }
        }
    })*/
})
