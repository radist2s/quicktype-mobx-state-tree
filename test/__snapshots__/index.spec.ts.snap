// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should import custom types module 1`] = `
"/* tslint:disable */
import {types} from '../src/types'

export enum Enum {
    First = \\"first\\",
    Second = \\"second\\",
    Third = \\"third\\",
}

export const Schema = types.model('Schema', {
    date: types.maybe(  types.Date),
    enum: types.maybe(  types.enumeration<Enum>('Enum', Object.values(Enum))),
    number: types.maybe(types.number),
    string: types.maybe(types.string),
})
"
`;

exports[`transform should load json snapshot 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": 6,
      "name": "Color",
      "options": Array [
        "Black",
        "Green",
      ],
      "position": 0,
      "variation": true,
      "visible": false,
    },
    Object {
      "id": 0,
      "name": "Size",
      "options": Array [
        "S",
        "M",
      ],
      "position": 0,
      "variation": true,
      "visible": true,
    },
  ],
  "backorders": "no",
  "button_text": "",
  "catalog_visibility": "visible",
  "categories": Array [
    Object {
      "id": 9,
      "name": "Clothing",
      "slug": "clothing",
    },
    Object {
      "id": 14,
      "name": "T-shirts",
      "slug": "t-shirts",
    },
  ],
  "cross_sell_ids": Array [],
  "date_created": "2017-03-23T12:03:12.000Z",
  "date_created_gmt": "2017-03-23T15:03:12.000Z",
  "date_on_sale_from": null,
  "date_on_sale_from_gmt": null,
  "date_on_sale_to": null,
  "date_on_sale_to_gmt": null,
  "default_attributes": Array [],
  "description": "<p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.</p>
",
  "dimensions": Object {
    "height": "",
    "length": "",
    "width": "",
  },
  "download_expiry": -1,
  "download_limit": -1,
  "downloadable": false,
  "downloads": Array [],
  "external_url": "",
  "featured": false,
  "id": 799,
  "images": Array [
    Object {
      "alt": "",
      "date_created": "2017-03-23T09:03:08.000Z",
      "date_created_gmt": "2017-03-23T15:03:08.000Z",
      "date_modified": "2017-03-23T09:03:08.000Z",
      "date_modified_gmt": "2017-03-23T15:03:08.000Z",
      "id": 795,
      "name": "",
      "src": "https://example.com/wp-content/uploads/2017/03/T_4_front-11.jpg",
    },
    Object {
      "alt": "",
      "date_created": "2017-03-23T09:03:09.000Z",
      "date_created_gmt": "2017-03-23T15:03:09.000Z",
      "date_modified": "2017-03-23T09:03:09.000Z",
      "date_modified_gmt": "2017-03-23T15:03:09.000Z",
      "id": 796,
      "name": "",
      "src": "https://example.com/wp-content/uploads/2017/03/T_4_back-10.jpg",
    },
    Object {
      "alt": "",
      "date_created": "2017-03-23T09:03:10.000Z",
      "date_created_gmt": "2017-03-23T15:03:10.000Z",
      "date_modified": "2017-03-23T09:03:10.000Z",
      "date_modified_gmt": "2017-03-23T15:03:10.000Z",
      "id": 797,
      "name": "",
      "src": "https://example.com/wp-content/uploads/2017/03/T_3_front-10.jpg",
    },
    Object {
      "alt": "",
      "date_created": "2017-03-23T09:03:11.000Z",
      "date_created_gmt": "2017-03-23T15:03:11.000Z",
      "date_modified": "2017-03-23T09:03:11.000Z",
      "date_modified_gmt": "2017-03-23T15:03:11.000Z",
      "id": 798,
      "name": "",
      "src": "https://example.com/wp-content/uploads/2017/03/T_3_back-10.jpg",
    },
  ],
  "manage_stock": false,
  "menu_order": 0,
  "meta_data": Array [],
  "name": "Ship Your Idea",
  "parent_id": 0,
  "purchase_note": "",
  "regular_price": "120.10",
  "reviews_allowed": true,
  "sale_price": "",
  "shipping_class": "",
  "short_description": "<p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.</p>
",
  "sku": "",
  "slug": "ship-your-idea-22",
  "sold_individually": false,
  "status": "publish",
  "stock_quantity": null,
  "stock_status": "instock",
  "tags": Array [],
  "tax_class": "",
  "tax_status": "taxable",
  "type": "variable",
  "upsell_ids": Array [],
  "virtual": false,
  "weight": "",
}
`;

exports[`transform should transform 1`] = `
"/* tslint:disable */
import {types} from 'mobx-state-tree'

export const Attributes = types.model('Attributes', {
    /**
     * Attribute ID.
     */
    id: types.maybe(types.number),
    /**
     * Attribute name.
     */
    name: types.maybe(types.string),
    /**
     * List of available term names of the attribute.
     */
    options: types.maybe(types.array(types.string)),
    /**
     * Attribute position.
     */
    position: types.maybe(types.number),
    /**
     * Define if the attribute can be used as variation.
     */
    variation: types.maybe(types.boolean),
    /**
     * Define if the attribute is visible on the \\"Additional information\\" tab in the product's
     * page.
     */
    visible: types.maybe(types.boolean),
})

/**
 * If managing stock, this controls if backorders are allowed.
 */
export enum Backorders {
    No = \\"no\\",
    Notify = \\"notify\\",
    Yes = \\"yes\\",
}

/**
 * Catalog visibility.
 */
export enum CatalogVisibility {
    Catalog = \\"catalog\\",
    Hidden = \\"hidden\\",
    Search = \\"search\\",
    Visible = \\"visible\\",
}

export const Categories = types.model('Categories', {
    /**
     * Category ID.
     */
    id: types.maybe(types.number),
    /**
     * Category name.
     */
    name: types.maybe(types.string),
    /**
     * Category slug.
     */
    slug: types.maybe(types.string),
})

export const DefaultAttributes = types.model('DefaultAttributes', {
    /**
     * Attribute ID.
     */
    id: types.maybe(types.number),
    /**
     * Attribute name.
     */
    name: types.maybe(types.string),
    /**
     * Selected attribute term name.
     */
    option: types.maybe(types.string),
})

export const Downloads = types.model('Downloads', {
    /**
     * File URL.
     */
    file: types.maybe(types.string),
    /**
     * File ID.
     */
    id: types.maybe(types.string),
    /**
     * File name.
     */
    name: types.maybe(types.string),
})

export const Images = types.model('Images', {
    /**
     * Image alternative text.
     */
    alt: types.maybe(types.string),
    /**
     * The date the image was created, in the site's timezone.
     */
    date_created: types.maybe(types.Date),
    /**
     * The date the image was created, as GMT.
     */
    date_created_gmt: types.maybe(types.Date),
    /**
     * The date the image was last modified, in the site's timezone.
     */
    date_modified: types.maybe(types.Date),
    /**
     * The date the image was last modified, as GMT.
     */
    date_modified_gmt: types.maybe(types.Date),
    /**
     * Image ID.
     */
    id: types.maybe(types.number),
    /**
     * Image name.
     */
    name: types.maybe(types.string),
    /**
     * Image URL.
     */
    src: types.maybe(types.string),
})

export const MetaData = types.model('MetaData', {
    /**
     * Meta ID.
     */
    id: types.maybe(types.number),
    /**
     * Meta key.
     */
    key: types.maybe(types.string),
    /**
     * Meta value.
     */
    value: types.maybe(types.union(types.array(types.frozen()), types.boolean, types.number, types.number, types.frozen(), types.null, types.string)),
})

/**
 * Product status (post status).
 */
export enum Status {
    Any = \\"any\\",
    AutoDraft = \\"auto-draft\\",
    Draft = \\"draft\\",
    Failed = \\"failed\\",
    Future = \\"future\\",
    InProgress = \\"in-progress\\",
    Inherit = \\"inherit\\",
    Pending = \\"pending\\",
    Private = \\"private\\",
    Publish = \\"publish\\",
    RequestCompleted = \\"request-completed\\",
    RequestConfirmed = \\"request-confirmed\\",
    RequestFailed = \\"request-failed\\",
    RequestPending = \\"request-pending\\",
    Trash = \\"trash\\",
    Vacation = \\"vacation\\",
    WcCancelled = \\"wc-cancelled\\",
    WcCompleted = \\"wc-completed\\",
    WcFailed = \\"wc-failed\\",
    WcOnHold = \\"wc-on-hold\\",
    WcPending = \\"wc-pending\\",
    WcProcessing = \\"wc-processing\\",
    WcRefunded = \\"wc-refunded\\",
    WcmActive = \\"wcm-active\\",
    WcmCancelled = \\"wcm-cancelled\\",
    WcmComplimentary = \\"wcm-complimentary\\",
    WcmDelayed = \\"wcm-delayed\\",
    WcmExpired = \\"wcm-expired\\",
    WcmPaused = \\"wcm-paused\\",
    WcmPending = \\"wcm-pending\\",
}

/**
 * Controls the stock status of the product.
 */
export enum StockStatus {
    Instock = \\"instock\\",
    Onbackorder = \\"onbackorder\\",
    Outofstock = \\"outofstock\\",
}

export const Tags = types.model('Tags', {
    /**
     * Tag ID.
     */
    id: types.maybe(types.number),
    /**
     * Tag name.
     */
    name: types.maybe(types.string),
    /**
     * Tag slug.
     */
    slug: types.maybe(types.string),
})

/**
 * Tax status.
 */
export enum TaxStatus {
    None = \\"none\\",
    Shipping = \\"shipping\\",
    Taxable = \\"taxable\\",
}

/**
 * Product type.
 */
export enum Type {
    External = \\"external\\",
    Grouped = \\"grouped\\",
    Simple = \\"simple\\",
    Variable = \\"variable\\",
}

export const Schema = types.model('Schema', {
    /**
     * List of attributes.
     */
    attributes: types.maybe(types.array(Attributes)),
    /**
     * If managing stock, this controls if backorders are allowed.
     */
    backorders: types.maybe(types.enumeration<Backorders>('Backorders', Object.values(Backorders))),
    /**
     * Product external button text. Only for external products.
     */
    button_text: types.maybe(types.string),
    /**
     * Catalog visibility.
     */
    catalog_visibility: types.maybe(types.enumeration<CatalogVisibility>('CatalogVisibility', Object.values(CatalogVisibility))),
    /**
     * List of categories.
     */
    categories: types.maybe(types.array(Categories)),
    /**
     * List of cross-sell products IDs.
     */
    cross_sell_ids: types.maybe(types.array(types.number)),
    /**
     * The date the product was created, in the site's timezone.
     */
    date_created: types.maybe(types.Date),
    /**
     * The date the product was created, as GMT.
     */
    date_created_gmt: types.maybe(types.Date),
    /**
     * Start date of sale price, in the site's timezone.
     */
    date_on_sale_from: types.maybe(types.Date),
    /**
     * Start date of sale price, as GMT.
     */
    date_on_sale_from_gmt: types.maybe(types.Date),
    /**
     * End date of sale price, in the site's timezone.
     */
    date_on_sale_to: types.maybe(types.Date),
    /**
     * End date of sale price, in the site's timezone.
     */
    date_on_sale_to_gmt: types.maybe(types.Date),
    /**
     * Defaults variation attributes.
     */
    default_attributes: types.maybe(types.array(DefaultAttributes)),
    /**
     * Product description.
     */
    description: types.maybe(types.string),
    /**
     * Product dimensions.
     */
    dimensions: types.maybe(types.frozen()),
    /**
     * Number of days until access to downloadable files expires.
     */
    download_expiry: types.maybe(types.number),
    /**
     * Number of times downloadable files can be downloaded after purchase.
     */
    download_limit: types.maybe(types.number),
    /**
     * If the product is downloadable.
     */
    downloadable: types.maybe(types.boolean),
    /**
     * List of downloadable files.
     */
    downloads: types.maybe(types.array(Downloads)),
    /**
     * Product external URL. Only for external products.
     */
    external_url: types.maybe(types.string),
    /**
     * Featured product.
     */
    featured: types.maybe(types.boolean),
    /**
     * Unique identifier for the resource.
     */
    id: types.maybe(types.number),
    /**
     * List of images.
     */
    images: types.maybe(types.array(Images)),
    /**
     * Stock management at product level.
     */
    manage_stock: types.maybe(types.boolean),
    /**
     * Menu order, used to custom sort products.
     */
    menu_order: types.maybe(types.number),
    /**
     * Meta data.
     */
    meta_data: types.maybe(types.array(MetaData)),
    /**
     * Product name.
     */
    name: types.maybe(types.string),
    /**
     * Product parent ID.
     */
    parent_id: types.maybe(types.number),
    /**
     * Optional note to send the customer after purchase.
     */
    purchase_note: types.maybe(types.string),
    /**
     * Product regular price.
     */
    regular_price: types.maybe(types.string),
    /**
     * Allow reviews.
     */
    reviews_allowed: types.maybe(types.boolean),
    /**
     * Product sale price.
     */
    sale_price: types.maybe(types.string),
    /**
     * Shipping class slug.
     */
    shipping_class: types.maybe(types.string),
    /**
     * Product short description.
     */
    short_description: types.maybe(types.string),
    /**
     * Unique identifier.
     */
    sku: types.maybe(types.string),
    /**
     * Product slug.
     */
    slug: types.maybe(types.string),
    /**
     * Allow one item to be bought in a single order.
     */
    sold_individually: types.maybe(types.boolean),
    /**
     * Product status (post status).
     */
    status: types.maybe(types.enumeration<Status>('Status', Object.values(Status))),
    /**
     * Stock quantity.
     */
    stock_quantity: types.maybe(types.union(types.number, types.null)),
    /**
     * Controls the stock status of the product.
     */
    stock_status: types.maybe(types.enumeration<StockStatus>('StockStatus', Object.values(StockStatus))),
    /**
     * List of tags.
     */
    tags: types.maybe(types.array(Tags)),
    /**
     * Tax class.
     */
    tax_class: types.maybe(types.string),
    /**
     * Tax status.
     */
    tax_status: types.maybe(types.enumeration<TaxStatus>('TaxStatus', Object.values(TaxStatus))),
    /**
     * Product type.
     */
    type: types.maybe(types.enumeration<Type>('Type', Object.values(Type))),
    /**
     * List of up-sell products IDs.
     */
    upsell_ids: types.maybe(types.array(types.number)),
    /**
     * If the product is virtual.
     */
    virtual: types.maybe(types.boolean),
    /**
     * Product weight (kg).
     */
    weight: types.maybe(types.string),
})
"
`;
